ophyd/signal.py

```diff
@@ -810,6 +810,45 @@ def limits(self):
         return (self._metadata['lower_ctrl_limit'],
                 self._metadata['upper_ctrl_limit'])
 
+    def get_or_timeout(self, *, as_string=None, form='time', 
+                timeout=None, **kwargs):
+        '''Get the readback value from PyEpics or raise TimeoutError
+
+        Parameters
+        ----------
+        as_string : bool, optional
+            Get a string representation of the value, defaults to as_string
+            from this signal, optional
+        timeout : float, optional
+            maximum time to wait for value to be received.
+            (default = 0.5 + log10(count) seconds)
+        form : {'time', 'ctrl'}
+            PV form to request
+
+        Returns
+        ------
+        info
+            a dictionary with the value and other metadata from PyEpics
+
+        Raises
+        ------
+        TimeoutError
+            When PyEpics realizes a timeout, it posts a warning rather
+            than raise an exception.  The message from PyEpics is
+            included in the exception message.
+        '''
+        warnings.filterwarnings("error")
```

Comment:
Can you use the context manager here (https://docs.python.org/3/library/warnings.html#warnings.catch_warnings)? 

If you don't want to use the context manager, can you use a `finally` block to reset the warnings?  If `self._read_pv.get_with_metadata` raises anything else but `UserWarning` the filter warnings will not be reset.

If the user has set any other filters (like we (should?) do in the test suite to ensure we have no warnings) this will clobber that and reset it back to the default behavior.  If we don't use the context manager you will also have to capture what the state used to be to reset it correctly.

Reply:
I was concerned about this note when using with a context manager:

> the context manager is modifying global state and therefore is not thread-safe.

Is that a problem here?  That's [why I chose the alternative method](https://github.com/bluesky/ophyd/issues/776#issuecomment-558849025).

Also, will a `finally` clause be called after an exception has been raised?

Reply:
Ah, the anything else condition.  Sure.

Reply:
How to undo the action of `warnings.filterwarnings("error")` in the `finally` clause? [`resetwarnings()`](https://docs.python.org/3.7/library/warnings.html#warnings.resetwarnings) looks like it resets too much.  Is there a way to get the state of warnings prior to calling `filterwarnings()`?

Reply:
A context would be easier but not thread-safe.

Reply:
Removed reference to `warnings` and removed `try..except..finally` clause due to thread-safe concerns.